docs/madr/decisions/056-meshexternalservice-and-zoneegress.md

```diff
@@ -0,0 +1,275 @@
+
+### Running traffic through sidecar only
+
+- all the problems of running through egress are not applicable (operational cost, extra networking hop, scalability etc.)
+- no single exit point for all the external traffic
+
+### Running traffic through sidecar and egress
+
+- we can mix and match the pros/cons of both approaches
+- it's harder to implement / explain how it's working
+
+### Configurable where it makes sense on both sidecar and egress, predefined where only one makes sense
+
+#### Table summary
+
+The table below shows a summary of policies and if it makes sense to configure them on:
+- sidecar
+- egress
+- on both
+
+and if it makes sense for the user to pick where.
+
+| Policy                    | On sidecar    | On egress              | On both | User configurable |
+|---------------------------|---------------|------------------------|---------|-------------------|
+| MeshAccessLog             | Yes (in `to`) | Yes (both `to`/`from`) | Yes     | Yes               |
+| MeshCircuitBreaker        | Yes           | Maybe                  | No      | No                |
+| MeshFaultInjection        | Maybe         | Yes                    | No      | No                |
+| MeshHealthCheck           | Yes           | Maybe                  | No      | No                |
+| MeshMetric                | Yes           | Yes                    | Yes     | Maybe             |
+| MeshProxyPatch            | Yes           | Yes                    | Yes     | Yes               |
+| MeshRateLimit             | Probably not  | Yes                    | No      | No                |
+| MeshRetry                 | Yes           | Maybe                  | No      | Maybe             |
+| MeshTimeout               | Yes           | Maybe                  | No      | No                |
+| MeshTrace                 | Yes           | Yes                    | Yes     | Maybe             |
+| MeshTrafficPermission     | Maybe         | Yes                    | Maybe   | Maybe             |
+| MeshLoadBalancingStrategy | Yes           | Yes                    | No      | Maybe             |
+| MeshTCPRoute              | Yes           | Yes                    | No      | No                |
+| MeshHTTPRoute             | Yes           | Yes                    | No      | No                |
+
+In paragraphs below I go deeper into each policy.
+**After PR review / discussion I will update this table to reflect the actual decision**.
+
+#### MeshAccessLog
+
+MeshAccessLog is not applied on egress at all, and we don't provide a way to configure it.
+I don't see any reason not to do this though.
+
+##### New behaviour
+
+#### MeshCircuitBreaker
+
+##### Old ExternalService behaviour
+
+Circuit Breaker is configured on the sidecar always.
+
+##### New behaviour
+
+If on egress
+- services from one zone will act as an aggregate - if a CB triggers by service-1 then service-2 will also see that result immediately
```

Comment:
can you elaborate on this one? Are you saying that CB configured on envoy cluster "service-1" is carried over to envoy cluster "service-2?

Reply:
what I'm thinking about is a scenario like this:

`serivce-a` consumes `es-1` through egress
`service-b` consumes `es-1` through egress

there is a bug in `service-a` that causes it to create lots of unnecessary connections, the circuitbreaker trips and starts rejecting connections for both `service-a` and `service-b` because on egress there is only one `es-1` cluster. We could theoretically work around this by having a cluster per service but that wouldn't scale for very long.

Reply:
Is this just a `to` vs `from` case?

In the `to` the CB is applied on the envoy cluster of `service-X`'s sidecar so A and B won't conflate.
In the `from` the CB is applied on the envoy cluster of `es-1` on each ZoneEgress therefore it applies to all clients.

I don't see how this is any different than regular dataplanes.

Reply:
> I don't see how this is any different than regular dataplanes.

I don't think it is and I don't think I claimed anywhere that it is.

Reply:
> there is a bug in service-a that causes it to create lots of unnecessary connections, the circuitbreaker trips and starts rejecting connections for both service-a and service-b because on egress there is only one es-1 cluster

I think that's acceptable. That's the cost of centralized access.

btw. in this example it would be very hard to hit this, because we do http2 between envoys

Reply:
I think we elaborated on the topic, feel free to re-open if you think it needs more info.